transports/quic/src/endpoint.rs

```diff
@@ -0,0 +1,471 @@
+    }
+
+    pub fn create_connection(
+        &self,
+        id: ConnectionHandle,
+    ) -> (ConnectionChannel<C>, mpsc::Sender<ConnectionEvent>) {
+        let (tx, rx) = mpsc::channel(12);
+        let channel = ConnectionChannel {
+            id,
+            tx: self.connection_tx.clone(),
+            rx,
+            port: self.port,
+            max_datagrams: self.max_datagrams,
+        };
+        (channel, tx)
+    }
+}
+
+type QuinnEndpointConfig<S> = quinn_proto::generic::EndpointConfig<S>;
+type QuinnEndpoint<S> = quinn_proto::generic::Endpoint<S>;
+
+pub struct EndpointConfig<C: Crypto> {
+    socket: UdpSocket,
+    endpoint: QuinnEndpoint<C::Session>,
+    port: u16,
+    crypto_config: Arc<CryptoConfig<C>>,
+    capabilities: UdpCapabilities,
+}
+
+impl<C: Crypto> EndpointConfig<C> {
+    pub fn new(mut config: QuicConfig<C>, addr: SocketAddr) -> Result<Self, QuicError> {
+        config.transport.max_concurrent_uni_streams(0)?;
+        config.transport.datagram_receive_buffer_size(None);
+        let transport = Arc::new(config.transport);
+
+        let crypto_config = Arc::new(CryptoConfig {
+            keypair: config.keypair,
+            psk: config.psk,
+            keylogger: config.keylogger,
+            transport: transport.clone(),
+        });
+
+        let mut server_config = ServerConfig::<C::Session>::default();
+        server_config.transport = transport;
+        server_config.crypto = C::new_server_config(&crypto_config);
+
+        let mut endpoint_config = QuinnEndpointConfig::default();
+        endpoint_config
+            .supported_versions(C::supported_quic_versions(), C::default_quic_version())?;
+
+        let socket = UdpSocket::bind(addr)?;
+        let port = socket.local_addr()?.port();
+        let endpoint = quinn_proto::generic::Endpoint::new(
+            Arc::new(endpoint_config),
+            Some(Arc::new(server_config)),
+        );
+        let capabilities = UdpSocket::capabilities()?;
+        Ok(Self {
+            socket,
+            endpoint,
+            port,
+            crypto_config,
+            capabilities,
+        })
+    }
+
+    pub fn spawn(self) -> TransportChannel<C>
+    where
+        <C::Session as Session>::ClientConfig: Send + Unpin,
+        <C::Session as Session>::HeaderKey: Unpin,
+        <C::Session as Session>::PacketKey: Unpin,
+    {
+        let (tx1, rx1) = mpsc::unbounded();
+        let (tx2, rx2) = mpsc::channel(1);
+        let transport = TransportChannel {
+            tx: tx1,
+            rx: rx2,
+            port: self.port,
+            ty: self.socket.socket_type(),
+        };
+        let endpoint = EndpointChannel {
+            tx: tx2,
+            rx: rx1,
+            port: self.port,
+            max_datagrams: self.capabilities.max_gso_segments,
+            connection_tx: transport.tx.clone(),
+        };
+        async_global_executor::spawn(Endpoint::new(endpoint, self)).detach();
```

Comment:
All of rust-libp2p is generic over the executor to be used. Before we introduce a protocol locked onto a single executor, what is the reasoning for not being generic over the executor?

Reply:
there is currently no api for a transport to use the swarm's executor